### PR TITLE
[Connector-V2] Add Kudu source and sink connector

### PR DESCRIPTION
**Pull request by 2013650523**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: plugin-mapping.properties [hunk 1]
```diff
@@ -102,6 +102,8 @@ seatunnel.sink.Clickhouse = connector-clickhouse
 seatunnel.sink.ClickhouseFile = connector-clickhouse
 seatunnel.source.Jdbc = connector-jdbc
 seatunnel.sink.Jdbc = connector-jdbc
+seatunnel.source.Kudu = connector-Kudu
```
Comment (CalvinKirs) [inline plugin-mapping.properties]: should be `seatunnel.source.Kudu = connector-kudu`

File: seatunnel-connectors-v2/connector-kudu/pom.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-kudu/pom.xml]: hi, you forget the license header.

File: seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduInputFormat.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        ArrayList<String> fieldNames = new ArrayList<>();
+        try {
+
+            for (int i = 0; i < columnSchemaList.size(); i++) {
+                fieldNames.add(columnSchemaList.get(i).getName());
+                seaTunnelDataTypes.add(KuduTypeMapper.mapping(columnSchemaList, i));
+            }
+        } catch (Exception e) {
+            logger .warn("get row type info exception", e);
+            throw new PrepareFailException("kudu", PluginType.SOURCE, e.toString());
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduInputFormat.java]: It's better to add an exception Message.

File: seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduOutputFormat.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+import java.nio.ByteBuffer;
+import java.sql.Timestamp;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A Kudu outputFormat
+ */
+public class KuduOutputFormat
+        implements Serializable {
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduOutputFormat.java:44]: Why not implement his function directly in `kuduSinkWriter`?
That can better support other kudu flush modes in the future.

File: seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduInputFormat.java (hunk 3 above)
Comment (2013650523) [reply seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduInputFormat.java]: throw new PrepareFailException("kudu", PluginType.SOURCE, e.toString());   I don't quite understand,There's an exception message in this one

File: seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduOutputFormat.java (hunk 4 above)
Comment (2013650523) [reply seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduOutputFormat.java:44]: sorry,I don't understand why the implementation of kuduSinkWriter will better support other kudu Flush modes. Can you tell me more about it? thank you

**Comment by EricJoy2048** [comment]
Can you change this pr title to `[Connector-V2]` ?

**Comment by 2013650523** [comment]
> Can you change this pr title to `[Connector-V2]` ?

ok,tks

File: seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/config/KuduSinkConfig.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                return OVERWRITE;
+            } else {
+                return APPEND;
+            }
+        }
+    }
+
+    public KuduSinkConfig(@NonNull Config pluginConfig) {
+
+        this.saveMode = StringUtils.isBlank(pluginConfig.getString(KUDU_SAVE_MODE)) ? SaveMode.APPEND : SaveMode.fromStr(pluginConfig.getString(KUDU_SAVE_MODE));
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/config/KuduSinkConfig.java]: There are some problem in this line

`pluginConfig.getString("xxxx")` will throw NullPointException when `xxxx` not in config file. So we need judge if `xxxx` is config file by `pluginConfig.hasPath("xxxx")`.

File: seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduInputFormat.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+     * Specifies the name of the table
+     */
+    public   String tableName;
+    public List<ColumnSchema>  getColumnsSchemas(){
+        List<ColumnSchema> columns = null;
+        try {
+            schema = kuduClient.openTable(tableName).getSchema();
+            keyColumn = schema.getPrimaryKeyColumns().get(0).getName();
+            columns =schema.getColumns();
+        } catch (KuduException e) {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/kuduclient/KuduInputFormat.java:82]: If the `KuduException` be catch,  this method will return `null`. But I found you didn't handle the `null` value when you use the return value of this method.

File: seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/sink/KuduSinkAggregatedCommitter.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class KuduSinkAggregatedCommitter implements SinkAggregatedCommitter<KuduCommitInfo, KuduAggregatedCommitInfo> {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kudu/src/main/java/org/apache/seatunnel/connectors/seatunnel/kudu/sink/KuduSinkAggregatedCommitter.java]: If the connector don't need store state and 2PC , You can only extends `AbstractSimpleSink` and that will be simply.

**Review by EricJoy2048 — COMMENTED** [review]
Some suggestions

**Comment by Hisoka-X** [comment]
Please resolve ci error. Thanks

**Comment by 2013650523** [comment]
Hi,@CalvinKirs I have finished the modification and provided the document, please help to review

**Review by EricJoy2048 — DISMISSED** [review]
+1 cc @CalvinKirs  @ashulin

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/ExceptionUtil.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            if (sw != null) {
+                try {
+                    sw.close();
+                } catch (IOException e1) {
+                    e1.printStackTrace();
+                }
+            }
+            if (pw != null) {
+                pw.close();
+            }
```
Comment (CalvinKirs) [inline seatunnel-common/src/main/java/org/apache/seatunnel/common/ExceptionUtil.java]: used `try-with-resources` to simplify this

**Review by Hisoka-X — APPROVED** [review]

**Pull request merged**